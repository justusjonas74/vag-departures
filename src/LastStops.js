import React, { Component } from 'react';
import './LastStops.css';

function StopListItem(props) {
  // Correct! There is no need to specify the key here:
  return <li className="list-inline-item"><button onClick={(e) => props.searchStop(e, props.stop.VGNKennung)} className="btn btn-outline-secondary btn-sm">{props.stop.Haltestellenname}</button></li>;
}

function StopList(props) {
  const lastStops = props.stops;
  const listItems = lastStops.map((stop) =>
    // Correct! Key should be specified inside the array.
    <StopListItem key={stop.VGNKennung.toString()}
              stop={stop}
              searchStop={props.searchStop}/>
  );
  return (
    <ul className="list-inline">
      {listItems}
    </ul>
  );
}



const formatOutput = (lastStops, searchStop) =>  {
      return (
         <div className="LastStopsContainer">
          <p>Your last searched stops:</p>
          <StopList stops={lastStops} searchStop={searchStop}/>
        </div>
      )
    }


class LastStops extends Component {
   
    
    render() {
      const lastStops = [...this.props.lastStops].reverse() //
     
      const output = lastStops.length > 0 ? formatOutput(lastStops,this.props.searchStop) : null
      return output
    }
}

export default LastStops;