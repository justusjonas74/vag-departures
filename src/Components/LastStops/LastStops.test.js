import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import LastStops from './LastStops';

describe('LastStops Component', () => {
  // BOILERPLATE
  let props;
  let mountedComponent;
  const lastStops = () => {
    if (!mountedComponent) {
      mountedComponent = mount(
        <LastStops {...props} />
      );
    }
    return mountedComponent;
  }

  beforeEach(() => {
    props = {
      lastStops: null,
      searchStop: null
    };
    mountedComponent = undefined;
  });
  
  it('should render without errors', ()=>{
     shallow(<LastStops />);
  })
  it('it should always render correctly', () => {
       const tree = renderer
          .create(<LastStops {...props} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
  });
  
  it('should return null if no props are present', ()=>{
    const component = lastStops().html()
    expect(component).toBeNull()
  })
  
  describe('with empty array as props.lastStops', ()=>{
    beforeEach(()=>{
      props.lastStops = []
    })
    
    it('should return null', ()=>{
      const component = lastStops().html()
      expect(component).toBeNull()
    })
  }) 
  
  describe('with props',()=>{
    
    beforeEach(()=>{
      props.lastStops = [{"Haltestellenname":"Bauernfeindstr. (Nürnberg)","VGNKennung":1550},{"Haltestellenname":"Hauptbahnhof (Nürnberg)","VGNKennung":510},{"Haltestellenname":"Plärrer (Nürnberg)","VGNKennung":704}]
    })
    
    it('should render correctly', ()=>{
        const tree = renderer
          .create(<LastStops {...props} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
    })
    
    it("always renders an unorderd div", () => {
      const component =  lastStops().find('div')
      expect(component.length).toEqual(1);
    });
     it("contains everything else that gets rendered", () => {
      const component = lastStops().find('div')
      const wrappingDiv = component.first();
      expect(wrappingDiv).toEqual(lastStops().children());
    });
    it('... with 3 li elements', ()=>{
      const listItems = lastStops().find('li')
      expect(listItems.length).toEqual(3)
    })
    
    it('... with 3 button elements', ()=>{
      const buttons = lastStops().find('button')
      expect(buttons.length).toEqual(3)
    })
    
    // TODO: test props.searchStop function
    
    describe('handle props.searchStop function correctly', ()=>{
      beforeEach(()=>{
        props.lastStops = [{"Haltestellenname":"Bauernfeindstr. (Nürnberg)","VGNKennung":1550}]

      })
      describe('Button', ()=>{
        it('should have a onClick function wich is equal to the mockFunction', ()=>{
          const mockFunction =  jest.fn()
          props.searchStop = mockFunction
          const component = lastStops()
          const button = component.find('button')
          button.simulate('click');
          expect(mockFunction.mock.calls.length).toEqual(1)
        })
        it('onClick should call the given function with stop id as param', ()=>{
          const mockFunction =  jest.fn((e,x)=>x)
          props.searchStop = mockFunction
          const component = lastStops()
          const button = component.find('button')
          button.simulate('click');
          expect(mockFunction.mock.results[0].value).toEqual(1550)
        })
      })
      
    })
  })
})

