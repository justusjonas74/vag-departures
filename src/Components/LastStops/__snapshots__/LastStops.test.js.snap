// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastStops Component it should always render correctly 1`] = `null`;

exports[`LastStops Component with props should render correctly 1`] = `
<div
  className="LastStopsContainer"
>
  <ul
    className="list-inline"
  >
    <li
      className="list-inline-item"
    >
      <button
        className="btn btn-outline-secondary btn-sm"
        onClick={[Function]}
      >
        Pl채rrer (N체rnberg)
      </button>
    </li>
    <li
      className="list-inline-item"
    >
      <button
        className="btn btn-outline-secondary btn-sm"
        onClick={[Function]}
      >
        Hauptbahnhof (N체rnberg)
      </button>
    </li>
    <li
      className="list-inline-item"
    >
      <button
        className="btn btn-outline-secondary btn-sm"
        onClick={[Function]}
      >
        Bauernfeindstr. (N체rnberg)
      </button>
    </li>
  </ul>
</div>
`;
